In [53]:
from sklearn.linear_model import LinearRegression
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline


In [54]:
data=pd.read_csv('../src/datasets/ecom-expense/Ecom Expense.csv')
data.head()

Transaction ID  Age    Items   Monthly Income  Transaction Time  Record  \
0         TXN001    42       10            7313        627.668127       5   
1         TXN002    24        8           17747        126.904567       3   
2         TXN003    47       11           22845        873.469701       2   
3         TXN004    50       11           18552        380.219428       7   
4         TXN005    60        2           14439        403.374223       2   

   Gender City Tier  Total Spend  
0  Female    Tier 1  4198.385084  
1  Female    Tier 2  4134.976648  
2    Male    Tier 2  5166.614455  
3  Female    Tier 1  7784.447676  
4  Female    Tier 2  3254.160485

In [5]:
# Crear Dummies variables
dummy_gender=pd.get_dummies(data['Gender'], prefix="Gender")
dummy_city_tier=pd.get_dummies(data['City Tier'], prefix='City')
dummy_city_tier
dummy_gender.head(10)

Gender_Female  Gender_Male
0              1            0
1              1            0
2              0            1
3              1            0
4              1            0
5              0            1
6              0            1
7              0            1
8              0            1
9              1            0

In [12]:
columns_names=data.columns.values.tolist()
# Unir columnas de Gender
df_new=data[columns_names].join(dummy_gender)


In [13]:
#Crear nuevo vector con nuevas columnas incluidas Gender con Tiers
columns_names2=df_new.columns.values.tolist()
df_new=df_new[columns_names2].join(dummy_city_tier)
df_new.head()

Transaction ID  Age    Items   Monthly Income  Transaction Time  Record  \
0         TXN001    42       10            7313        627.668127       5   
1         TXN002    24        8           17747        126.904567       3   
2         TXN003    47       11           22845        873.469701       2   
3         TXN004    50       11           18552        380.219428       7   
4         TXN005    60        2           14439        403.374223       2   

   Gender City Tier  Total Spend  Gender_Female  Gender_Male  City_Tier 1  \
0  Female    Tier 1  4198.385084              1            0            1   
1  Female    Tier 2  4134.976648              1            0            0   
2    Male    Tier 2  5166.614455              0            1            0   
3  Female    Tier 1  7784.447676              1            0            1   
4  Female    Tier 2  3254.160485              1            0            0   

   City_Tier 2  City_Tier 3  
0            0            0  
1            1            0  
2            1            0  
3            0            0  
4            1            0

In [16]:
#DEfinir las variables que dependera el modelo
feature_cols=['Monthly Income', 'Transaction Time', 
              'Gender_Female', 'Gender_Male', 
              'City_Tier 1', 'City_Tier 2', 'City_Tier 3',
              'Record' 
             ]


In [22]:
X=df_new[feature_cols]
Y=df_new['Total Spend']


In [23]:
lm= LinearRegression()
lm.fit(X,Y)


LinearRegression()

In [25]:
print(lm.intercept_)
print(lm.coef_)

-79.4171303013718
[ 1.47538980e-01  1.54946125e-01 -1.31025013e+02  1.31025013e+02
  7.67643260e+01  5.51389743e+01 -1.31903300e+02  7.72233446e+02]


#### Se debera juntar los elementos par saber como queda el modelo par eso se usa ZIP

In [18]:
a=list(zip(feature_cols, lm.coef_))
a
#Lista de tuplas para saber cuanto vale cada modelo

[('Monthly Income', 0.14753898049205738),
 ('Transaction Time', 0.15494612549589634),
 ('Gender_Female', -131.02501325554624),
 ('Gender_Male', 131.02501325554607),
 ('City_Tier 1', 76.76432601049513),
 ('City_Tier 2', 55.1389743092325),
 ('City_Tier 3', -131.9033003197277),
 ('Record', 772.2334457445645)]

In [26]:
# R^2 SCORE
lm.score(X,Y)

0.9179923586131016

## El modelo puede ser escrito como

* *Total  Spend*=  -79.4171303013718  +  0.14753898049205738  _'Monthly Income'_ +  0.15494612549589634 * _'Transaction Time'_ --131.02501325554624* 'Gender_Female' + 131.02501325554607* 'Gender_Male' + 76.76432601049513 * 'City_Tier 1' + 55.1389743092325 * 'City_Tier 2' -131.9033003197277 * 'City_Tier 3' +772.2334457445645 * 'Record'

In [32]:
df_new['prediction']=  (-79.4171303013718)  +  df_new['Monthly Income'] * 0.14753898049205738 +0.15494612549589634 * df_new['Transaction Time'] +(-131.02501325554624) * df_new['Gender_Female']+  131.02501325554607 * df_new['Gender_Male'] + 76.76432601049513 * df_new['City_Tier 1'] + 55.1389743092325 *df_new['City_Tier 2'] + (-131.9033003197277) * df_new['City_Tier 3'] + 772.2334457445645 * df_new['Record']
df_new.head()

Transaction ID  Age    Items   Monthly Income  Transaction Time  Record  \
0         TXN001    42       10            7313        627.668127       5   
1         TXN002    24        8           17747        126.904567       3   
2         TXN003    47       11           22845        873.469701       2   
3         TXN004    50       11           18552        380.219428       7   
4         TXN005    60        2           14439        403.374223       2   

   Gender City Tier  Total Spend  Gender_Female  Gender_Male  City_Tier 1  \
0  Female    Tier 1  4198.385084              1            0            1   
1  Female    Tier 2  4134.976648              1            0            0   
2    Male    Tier 2  5166.614455              0            1            0   
3  Female    Tier 1  7784.447676              1            0            1   
4  Female    Tier 2  3254.160485              1            0            0   

   City_Tier 2  City_Tier 3   prediction  
0            0            0  4903.696720  
1            1            0  4799.434826  
2            1            0  5157.082504  
3            0            0  8068.012996  
4            1            0  3581.980335

In [37]:
SSD=np.sum((df_new['prediction']-df_new['Total Spend'])**2)
SSD

1517733985.340816

In [45]:
RSE=np.sqrt(SSD/(len(df_new)-len(feature_cols)-1))
RSE

803.1318809818165

In [40]:
sales_mean=np.mean(df_new['Total Spend'])
sales_mean

6163.176415976715

In [46]:
error=RSE/sales_mean
error

0.1303113568029416

In [47]:
print("ERROR ", error*100, "%")

ERROR  13.03113568029416 %


## EL modelo no es cierto ya que depende de 2 dummy variables de Genero y 3 de City. es decir que dependera de 6 modelos reales
* Ya que las Dummy variables son 1 si aplica y 0 si no aplica por lo que todo da lo mismo o da 0.


## El modelo puede ser escrito como
### SI es Homb re y vive en City Tier 1
## EJEMPLOS DE MODELOS Definitivos
## 
#### Teniendo en cuenta lsa Variables Dummy por lo cual no requiere esa variable y se reduce el modelo
#### Si es Gender_Male=1 City_Tier=1  
* *Total  Spend*=  -79.4171303013718  +  0.14753898049205738  _'Monthly Income'_ +  0.15494612549589634 * _'Transaction Time'_  + 131.02501325554607*  + 76.76432601049513 *   +772.2334457445645 * 'Record'

#### Si es Gender_Female=1 City_Tier=1  
* *Total  Spend*=  -79.4171303013718  +  0.14753898049205738  _'Monthly Income'_ +  0.15494612549589634 * _'Transaction Time'_ --131.02501325554624* 'Gender_Female' + 131.02501325554607*  + 76.76432601049513 *   +772.2334457445645 * 'Record'


In [50]:
## El modelo puede ser escrito como

#* *Total  Spend*=  -79.4171303013718  +  0.14753898049205738  _'Monthly Income'_ +  0.15494612549589634 * _'Transaction Time'_ --131.02501325554624* 'Gender_Female' + 131.02501325554607* 'Gender_Male' + 76.76432601049513 * 'City_Tier 1' + 55.1389743092325 * 'City_Tier 2' -131.9033003197277 * 'City_Tier 3' +772.2334457445645 * 'Record'

In [64]:
data=pd.read_csv('../src/datasets/ecom-expense/Ecom Expense.csv')
data.head()
#Eliminar la columna Solo queda Male, 

dummy_gender=pd.get_dummies(df['Gender'], prefix='Gender').iloc[:,1:]
dummy_gender
#Elimina la columna 0 Solo queda City 2 o 3
dummy_city_tier=pd.get_dummies(df['City Tier'], prefix='City').iloc[:,1:]
dummy_city_tier

City_Tier 2  City_Tier 3
0               0            0
1               1            0
2               1            0
3               0            0
4               1            0
...           ...          ...
2357            1            0
2358            1            0
2359            0            1
2360            0            0
2361            0            0

[2362 rows x 2 columns]

In [73]:
column_names=df.columns.values.tolist()
df_new=df[column_names].join(dummy_gender)
column_names=df_new.columns.values.tolist()
df_new=df_new[column_names].join(dummy_city_tier)
df_new

Transaction ID  Age    Items   Monthly Income  Transaction Time  Record  \
0            TXN001    42       10            7313        627.668127       5   
1            TXN002    24        8           17747        126.904567       3   
2            TXN003    47       11           22845        873.469701       2   
3            TXN004    50       11           18552        380.219428       7   
4            TXN005    60        2           14439        403.374223       2   
...             ...   ...      ...             ...               ...     ...   
2357        TXN2358    50        7            5705        460.157207       3   
2358        TXN2359    35       11           11202        851.924751       8   
2359        TXN2360    27        5           21335        435.145358       8   
2360        TXN2361    45       12           19294        658.439838       7   
2361        TXN2362    46        7            2855        560.514341       8   

      Gender City Tier  Total Spend  Gender_Male  City_Tier 2  City_Tier 3  
0     Female    Tier 1  4198.385084            0            0            0  
1     Female    Tier 2  4134.976648            0            1            0  
2       Male    Tier 2  5166.614455            1            1            0  
3     Female    Tier 1  7784.447676            0            0            0  
4     Female    Tier 2  3254.160485            0            1            0  
...      ...       ...          ...          ...          ...          ...  
2357    Male    Tier 2  2909.619546            1            1            0  
2358    Male    Tier 2  7968.633136            1            1            0  
2359  Female    Tier 3  8816.406448            0            0            1  
2360  Female    Tier 1  7915.595856            0            0            0  
2361    Male    Tier 1  6832.401686            1            0            0  

[2362 rows x 12 columns]

In [79]:
feature_cols=['Monthly Income', 'Transaction Time', 'Gender_Male', 'City_Tier 2', 'City_Tier 3','Record']
X= df_new[feature_cols]
Y= df_new['Total Spend']
lm= LinearRegression()
lm.fit(X,Y)


LinearRegression()

In [81]:
print(lm.intercept_)

-133.67781754642238


In [84]:
list(zip(feature_cols, lm.coef_))

[('Monthly Income', 0.14753898049205744),
 ('Transaction Time', 0.1549461254959002),
 ('Gender_Male', 262.0500265110948),
 ('City_Tier 2', -21.62535170126276),
 ('City_Tier 3', -208.66762633022296),
 ('Record', 772.2334457445636)]

In [86]:
lm.score(X,Y)

0.9179923586131016

#### Coeficientes tras mascarar las variables Dummy que no se necesitan 

* [('Monthly Income', 0.14753898049205744),
 ('Transaction Time', 0.1549461254959002),
 ('Gender_Male', 262.0500265110948),
 ('City_Tier 2', -21.62535170126276),
 ('City_Tier 3', -208.66762633022296),
 ('Record', 772.2334457445636)]

# Transformación de variables para obtener modelo : relacion no lineal

### *Se consigue con una matriz de scater plot para saber como se distribuyen